### ISSUE TITLE
Messages not making it to SQL Server

### ISSUE THREAD

**Issue by mkoziel2000** [rebus-org/Rebus#810]
I'm running into an intermittent issue where some messages will not reach the message queue table in SQL Server.  I diagnosed the problem and I think it has to do with how Rebus is handling transactions.  Every once in a while, the ExecuteNonQueryAsync() call in the Rebus.SqlServer module will return a -1 instead of 1 indicating that the message wasn't written to the table in SQLServer (As a side note, it would be nice to get that return value evaluated and at least log it if its anything other than the intended value.)

I have attached a sample program that demonstrates the issue where I tried to line up the Rebus config and message pub/sub handlers as close as I could to what we have deployed.  The sample program should end with a statement saying that 41 messages where received.  It tends to never get this far because at least one message never gets on the queue.

I also have some commented code in the MySaga.cs, whereby I turn on RebusTransactionScope() for one of the publish() calls and it all works as expected.  I'd really like to know why this is and what the recommended pathway forward would be.  When I tried adding RebusTransactionScope around all the publish calls, it seems to get really flaky as to whether any of the messages make it onto the Q.  Its only when I add transaction scope to a few of the Publish calls does it actually work consistently.  Are we dealing with a bug or is there a best practices that we can follow to address this scenario?
[RebusSqlProblem.zip](https://github.com/rebus-org/Rebus/files/3316550/RebusSqlProblem.zip)

**Comment by mookid8000** [comment]
Hmm, you're postulating a lot of different things here.... for example, I have no idea what you mean by this:

> (...) I think it has to do with how Rebus is handling transactions (...)

?

> (...) Every once in a while, the ExecuteNonQueryAsync() call in the Rebus.SqlServer module will return a -1 instead of 1 indicating that the message wasn't written to the table in SQLServer (...)

Do you have any documentation of that behaviour? As far as I can tell, `ExecuteNonQuery` (and its async equivalent, `ExecuteNonQueryAsync`) should return -1 if you're invoking a stored procedure with `SET NOCOUNT ON`. I've never heard that it should be necessary to check the return value from `ExecuteNonQuery` when doing an INSERT – either it gets successfully inserted, or it doesn't (with the caveat that the SQL transaction might get rolled back).

## Here's an explanation of the problems in your code

The parallelism settings here
```csharp
                    o.SetNumberOfWorkers(5);
                    o.SetMaxParallelism(5);
```
configure Rebus to be able to process 5 messages concurrently.

The fact that you initially publish 20 events that ALL HIT THE SAME SAGA INSTANCE means that there's A LOT of contention around that instance.

Contention around saga instances means rollbacks, because the saga data update will result in a `ConcurrencyException`.

The fact that you

* track handled messages in a static in-mem dictionary
* publish `Level2Msg` AFTER registering successfully having processed `Level1Msg`
* in many cases end up rolling back the publish operation for `Level2Msg`

will simply result in having registered many successfully handled messages, even though that was not the case.

## How can you solve it?

### Avoid race conditions by setting parallelism to 1

Either reduce the parallelism to 1 (i.e. no concurrency):
```csharp
                    o.SetNumberOfWorkers(1);
                    o.SetMaxParallelism(1);
```
then it works flawlessly:
![image](https://user-images.githubusercontent.com/32968/59962288-9a604c00-94e3-11e9-92af-b5df49f41d30.png)

### Let Rebus handle concurrency

Let's track successful processing of messages by some other means, i.e. one that actually makes sense – could for example be in the saga: So I add a little counter in the saga data:
```csharp
public class MySagaData : IIdempotentSagaData
{
    public int sagaId { get; set; }
    public Guid Id { get; set; }
    public int Revision { get; set; }
    public IdempotencyData IdempotencyData { get; set; }
    public int MessageCount { get; set; } //< this one
}
```
and then I disable all the in-mem double-processing checks:

![image](https://user-images.githubusercontent.com/32968/59962358-9aad1700-94e4-11e9-9a47-d94c41e36a38.png)

and then I increment `MessageCount` in the saga data in every handler:
```csharp
                Data.MessageCount++;
```

When I run the app, LOTS of concurrency exceptions will occur, and some of the events will end up in the error queue:

![image](https://user-images.githubusercontent.com/32968/59962386-098a7000-94e5-11e9-8892-4dc8a1988218.png)

So I move those back into their input queue:
```sql
begin transaction
	insert into queue (priority, expiration, visible, headers, body) select priority, expiration, visible, headers, body from errq
	delete from errq
commit transaction
```
and then the world continues as expected, and we end up here:

![image](https://user-images.githubusercontent.com/32968/59962344-6fc2c300-94e4-11e9-8e10-d429030060af.png)

![image](https://user-images.githubusercontent.com/32968/59962353-85d08380-94e4-11e9-8dcf-c417b03df0f9.png)

### Write a conflict handler

The last way to handle saga concurrency issues, is to provide a conflict handler – in `MySaga` you can override `ResolveConflict`, which makes it possible for you to provide code that MERGES the state of two conflicting saga instances.

To be able to merge two saga instances, I need to track more that just the message count – therefore, I now change the saga data to be able to track IDs of handled messages

![image](https://user-images.githubusercontent.com/32968/59962523-ecef3780-94e6-11e9-8152-dfbf9ade2d2c.png)

and then I change the logic to

```csharp
                Data.MessageIds.Add(_messageContext.Headers.GetValue(Headers.MessageId));
                Data.MessageCount = Data.MessageIds.Count;
```

in all the places where I simply incremented the message count before.

That makes it possible for me to write this conflict handler that merges the state of two saga data instances:

```csharp
        protected override async Task ResolveConflict(MySagaData otherSagaData)
        {
            // the other saga data comes from the handling of 1 other event,
            // so we just increment our counter here - REAL logic would probably
            // be more involved
            Data.MessageIds = new HashSet<string>(Data.MessageIds.Concat(otherSagaData.MessageIds));
            Data.MessageCount = Data.MessageIds.Count;
        }
```

With that change, the execution of your program goes like this:

![image](https://user-images.githubusercontent.com/32968/59962443-c54b9f80-94e5-11e9-95c1-b8748c0ea0fa.png)

and the saga state ends up like this:

![image](https://user-images.githubusercontent.com/32968/59962549-52432880-94e7-11e9-8f5c-4aa3d88b881e.png)


### Conclusion

It doesn't seem like you've found any issues with concurrency or transactions in Rebus. 😄 Thanks for raising the issue though.

**Comment by mkoziel2000** [comment]
Wow...great explanation of options.  I actually do have a conflict resolver in our production code...I didn't add it in the example as I didn't think it was pertinent to the problem since I was able to replicate the original issue.  Maybe the key takeaway here (for me) is your technique on idempotentency.  I did not know that there was a message id in the header that we could use to arrest duplicates.  Good info to have.  Let me take your suggestions and try them out.  Hopefully this gets us back on the straight and narrow.

As for the return value from ExecuteNonQuery(), MSDN literature states that the number returned represents the number of records inserted/updated/deleted and that -1 represents something abnormal, like a rollback.  I didn't see a NO COUNT being applied to the INSERT statement built so I assumed it would be returning 1 each time.  When I added a log statement to check the value within the InnerSend() routine, it was sending back -1 for the instances where the messages where not getting to the DB...this is what originally sent me down the transaction route since (according to the doc) -1 could represent a rollback being performed.

https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlcommand.executenonquery?view=netframework-4.8

BTW, you got a great tooling library here and your support of it is crazy good.  Tx

-Mike

**Comment by mookid8000** [comment]
You're right that the docs state that

> (...) If a rollback occurs, the return value is also -1 (...)

but I can't see how an INSERT can result in a rollback?

It's funny though, that you've been able to log examples of the value being -1.... It never occurred to me that it should be necessary to check the return value, because you get an exception if the insert does not succeed.

I'm actually fairly convinced that it shouldn't be necessary to check the return value – see e.g. here how NServiceBus sends a message using its SQL transport: [TableBasedQueue.cs](https://github.com/Particular/NServiceBus.SqlServer/blob/master/src/NServiceBus.SqlServer/Queuing/TableBasedQueue.cs#L86-L91)

**Comment by mkoziel2000** [comment]
ok...I played with the options you laid out.  What I'm finding is that the message INSERT will fail from time to time - maybe from a DB issue, maybe from SagaData rollback as you stated.  Either way, the "at least once" message delivery is kicking in (the good).  My idempotency logic is essentially detecting that I'm getting this message a second time and therefore I'm just exiting to indicate that I've already done the work (the bad).  In reality, because the message handler also publishes out its own messages to get picked up by additional handlers...those downstream publications could have been what caused the message processing to fail in the first place.  This behavior + my idempotency logic is what is causing messages not to get posted and processed.  The failed INSERT was a red herring as I was focusing on the failure that kicked it all off by itself and not the recovery process as a whole.

Long story short: 

Before:
Handle(msg) 
{
   if (!AlreadyProcessed)
   {
      //do work here
      _bus.Publish(addtlMsg);
   }
}


After:
Handle(msg) 
{
   if (!AlreadyProcessed)
   {
      //do work here
   }
   // Always publish 
   _bus.Publish(addtlMsg);
}
